File: src/ros2/src/bob_simulate/bob_simulate/simulation_overlay_provider_node.py
import traceback as tb
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSDurabilityPolicy, QoSHistoryPolicy
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
from typing import List
from sensor_msgs.msg import Image
from bob_shared.node_runner import NodeRunner
from .object_simulator import MovingCircle

class SimulationOverlayProviderNode(Node):

  def __init__(self, subscriber_qos_profile: QoSProfile, publisher_qos_profile: QoSProfile):
    super().__init__('bob_overlayed_video_provider')

    self.br = CvBridge()

    self.declare_parameters(
        namespace="",
        parameters=[                
            ("height", 1080),
            ("width", 1920),
            ("target_object_diameter", 5)])

    # grab parameters provided
    self.height = self.get_parameter('height').value
    self.width = self.get_parameter('width').value
    self.target_object_diameter = self.get_parameter('target_object_diameter').value

    self.moving_circles = [MovingCircle(self.width, self.height, 0) for i in range(4)]

    # setup services, publishers and subscribers
    self.sub_camera = self.create_subscription(Image, 'bob/simulation/input_frame', self.camera_callback, subscriber_qos_profile)
    self.pub_synthetic_frame = self.create_publisher(Image, 'bob/simulation/output_frame', publisher_qos_profile)

    self.get_logger().info(f'{self.get_name()} node is up and running.')

  def camera_callback(self, msg_image:Image):

    if not msg_image is None:

      try:
        input_frame = self.br.imgmsg_to_cv2(msg_image)
        for circle in self.moving_circles:
            circle.move()
            circle.draw(input_frame, (10, 10, 20))
        frame_synthetic_msg = self.br.cv2_to_imgmsg(input_frame, encoding=msg_image.encoding)
        frame_synthetic_msg.header = msg_image.header
        self.pub_synthetic_frame.publish(frame_synthetic_msg)        
      except Exception as e:
        self.get_logger().error(f"Exception during frame overlay simulation. Error: {e}.")
        self.get_logger().error(tb.format_exc())

def main(args=None):

  rclpy.init(args=args)

  subscriber_qos_profile = QoSProfile(depth=10)
  subscriber_qos_profile.reliability = QoSReliabilityPolicy.BEST_EFFORT
  subscriber_qos_profile.durability = QoSDurabilityPolicy.VOLATILE
  subscriber_qos_profile.history = QoSHistoryPolicy.KEEP_LAST

  publisher_qos_profile = QoSProfile(depth=10)
  publisher_qos_profile.reliability = QoSReliabilityPolicy.BEST_EFFORT
  publisher_qos_profile.durability = QoSDurabilityPolicy.VOLATILE
  publisher_qos_profile.history = QoSHistoryPolicy.KEEP_LAST

  node = SimulationOverlayProviderNode(subscriber_qos_profile, publisher_qos_profile)

  runner = NodeRunner(node)
  runner.run()


if __name__ == '__main__':
  main()

File: src/ros2/src/bob_simulator/src/simulation_overlay_provider_node.cpp
#include <chrono>
#include <memory>
#include <vector>

#include <rclcpp/rclcpp.hpp>
#include <rclcpp_components/register_node_macro.hpp>

#include <cv_bridge/cv_bridge.hpp>
#include <sensor_msgs/msg/image.hpp>

#include "parameter_node.hpp"
#include <visibility_control.h>

#include "circle_frame_generator.cpp" 
#include "image_utils.hpp"

class SimulationOverlayProviderNode 
    : public ParameterNode 
{
public:
    COMPOSITION_PUBLIC
    explicit SimulationOverlayProviderNode(const rclcpp::NodeOptions & options) 
    : ParameterNode("simulation_overlay_provider_node", options) 
    {
        rclcpp::QoS subscriber_qos_profile(10);
        subscriber_qos_profile.reliability(rclcpp::ReliabilityPolicy::BestEffort);
        subscriber_qos_profile.durability(rclcpp::DurabilityPolicy::Volatile);
        subscriber_qos_profile.history(rclcpp::HistoryPolicy::KeepLast);

        rclcpp::QoS publisher_qos_profile(10);
        publisher_qos_profile.reliability(rclcpp::ReliabilityPolicy::BestEffort);
        publisher_qos_profile.durability(rclcpp::DurabilityPolicy::Volatile);
        publisher_qos_profile.history(rclcpp::HistoryPolicy::KeepLast);

        declare_node_parameters();

        std::pair<int, int> default_size_range(2, 10); 
        std::pair<int, int> default_step_range(5, 30);  

        for(int i = 0; i < num_simulated_objects_; i++) {
            moving_circles_.emplace_back(width_, height_, 0.5, default_size_range, default_step_range);
        }

        sub_camera_ = this->create_subscription<sensor_msgs::msg::Image>(
            "bob/simulation/input_frame",
            subscriber_qos_profile,
            std::bind(&SimulationOverlayProviderNode::camera_callback, this, std::placeholders::_1)
        );

        pub_synthetic_frame_ = this->create_publisher<sensor_msgs::msg::Image>("bob/simulation/output_frame", publisher_qos_profile);
    }

    void declare_node_parameters()
    {
        std::vector<ParameterNode::ActionParam> params = {
            ParameterNode::ActionParam(
                rclcpp::Parameter("height", 1080), 
                [this](const rclcpp::Parameter& param) {height_ = param.as_int();}
            ),
            ParameterNode::ActionParam(
                rclcpp::Parameter("width", 1920), 
                [this](const rclcpp::Parameter& param) {width_ = param.as_int();}
            ),
            ParameterNode::ActionParam(
                rclcpp::Parameter("num_objects", 5), 
                [this](const rclcpp::Parameter& param) {num_simulated_objects_ = param.as_int();}
            ),
        };
        add_action_parameters(params);
    }

private:
    int height_, width_, num_simulated_objects_;
    std::vector<MovingCircle> moving_circles_;
    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr sub_camera_;
    rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr pub_synthetic_frame_;

    void camera_callback(const sensor_msgs::msg::Image::SharedPtr msg_image) {

        if (msg_image->data.empty()) {
            RCLCPP_WARN(get_logger(), "Received an empty image frame. Skipping processing.");
            return;
        }

        try {
            
            cv::Mat input_frame;
            ImageUtils::convert_image_msg(msg_image, input_frame, false);

            for(auto &circle : moving_circles_) {
                circle.move();
                circle.draw(input_frame, cv::Scalar(10, 10, 20));
            }
            auto frame_synthetic_msg = cv_bridge::CvImage(msg_image->header, input_frame.channels() == 1 ? sensor_msgs::image_encodings::MONO8 : sensor_msgs::image_encodings::BGR8, input_frame).toImageMsg();
            pub_synthetic_frame_->publish(*frame_synthetic_msg);


        } catch(const std::exception& e) {
            RCLCPP_ERROR(get_logger(), "Exception during frame overlay simulation: %s", e.what());
        }
    }
};

int main(int argc, char **argv) 
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SimulationOverlayProviderNode>(rclcpp::NodeOptions()));
    rclcpp::shutdown();
    return 0;
}


RCLCPP_COMPONENTS_REGISTER_NODE(SimulationOverlayProviderNode)